rust/src/tests/chown.rs

```diff
@@ -0,0 +1,1074 @@
+
+    assert!(lchown(&link, Some(user.uid), Some(group.gid)).is_ok());
+
+    lchmod(&link, mode).unwrap();
+
+    let file_stat = symlink_metadata(&link).unwrap();
+    assert_eq!(file_stat.mode() as mode_t & ALLPERMS_SETBITS, mode.bits());
+    assert_eq!(file_stat.uid() as uid_t, user.uid.as_raw());
+    assert_eq!(file_stat.gid() as gid_t, group.gid.as_raw());
+
+    assert!(lchown(&link, Some(other_user.uid), Some(other_group.gid)).is_ok());
+    let file_stat = symlink_metadata(&link).unwrap();
+    let actual_mode = file_stat.mode() as mode_t & ALLPERMS_SETBITS;
+    assert!(actual_mode == mode.bits() || actual_mode == mode_without_setbits.bits());
+    assert_eq!(file_stat.uid() as uid_t, other_user.uid.as_raw());
+    assert_eq!(file_stat.gid() as gid_t, other_group.gid.as_raw());
+
+    lchmod(&link, mode).unwrap();
+
+    assert!(lchown(&link, Some(root.uid), Some(root.gid)).is_ok());
+    let file_stat = symlink_metadata(&link).unwrap();
+    let actual_mode = file_stat.mode() as mode_t & ALLPERMS_SETBITS;
+    assert!(actual_mode == mode.bits() || actual_mode == mode_without_setbits.bits());
+    assert_eq!(file_stat.uid() as uid_t, root.uid.as_raw());
+    assert_eq!(file_stat.gid() as gid_t, root.gid.as_raw());
+}
+
+crate::test_case! {
+    /// when non-super-user calls chown(2) successfully, set-uid and set-gid bits may
+    /// be removed, except when both uid and gid are equal to -1.
+    user_remove_suid_sgid, serialized, root => [Regular, Dir, Fifo, Block, Char, Socket]
+}
+fn user_remove_suid_sgid(ctx: &mut SerializedTestContext, ft: FileType) {
+    let file = ctx.create(ft.clone()).unwrap();
+    let (user, group) = ctx.get_new_entry();
+    let other_group = ctx.get_new_group();
+
+    let mode = Mode::from_bits_truncate(0o555) | Mode::S_ISUID | Mode::S_ISGID;
+    let mode_without_setbits = mode & !Mode::S_ISUID & !Mode::S_ISGID;
+
+    assert!(chown(&file, Some(user.uid), Some(group.gid)).is_ok());
+    chmod(&file, mode).unwrap();
+
+    let file_stat = metadata(&file).unwrap();
+    assert_eq!(file_stat.mode() as mode_t & ALLPERMS_SETBITS, mode.bits());
+    assert_eq!(file_stat.uid() as uid_t, user.uid.as_raw());
+    assert_eq!(file_stat.gid() as gid_t, group.gid.as_raw());
+
+    ctx.as_user(&user, Some(&[group.gid, other_group.gid]), || {
+        assert!(chown(&file, Some(user.uid), Some(other_group.gid)).is_ok());
+    });
+
+    let file_stat = metadata(&file).unwrap();
+    // TODO: Linux doesn't clear the SGID/SUID bits for directories, despite the description noted
+    // Linux makes a destinction for behavior when an executable file vs a
+    // non-executable file. From chmod(2):
+    //
+    //   When the owner or group of an executable file are changed by an
+    //   unprivileged user the S_ISUID and S_ISGID mode bits are cleared.
+    //
+    // I believe in this particular case, the behavior's bugged.
+
+    if cfg!(not(target_os = "linux")) || ft != FileType::Dir {
```

Comment:
Hmm, I'm unsure on how to implement this. Maybe adding support first for skipping a list of tests defined in the configuration file and define the default one for each OS? Because this isn't the only bad behavior encoded in the test suite.

Reply:
In this case we could do
```rust
#[cfg(not(target_os = "linux"))]
crate::test_case! {
    /// something something something
    user_remove_suid_sgid, serialized, root => [Regular, Dir, Fifo, Block, Char, Socket]
}
#[cfg(target_os = "linux")]
crate::test_case! {
    /// something something something
    user_remove_suid_sgid, serialized, root => [Regular, Fifo, Block, Char, Socket]
}
```
More generally I think you're right that the configuration file will need a way to specify certain tests to skip.  Or, in ATF parlance, to "expect_fail".